std/math/big/int.zig

```diff
@@ -0,0 +1,2023 @@
+                .limbs = block: {
+                    var limbs = allocator.alloc(Limb, Int.default_capacity) catch unreachable;
+                    limbs[0] = 0;
+                    break :block limbs;
+                },
+                .len = 1,
+            };
+            s.set(bn) catch unreachable;
+            return s;
+        },
+    }
+}
+
+pub const Int = struct {
+    allocator: *Allocator,
+    positive: bool,
+    //  - little-endian ordered
+    //  - len >= 1 always
+    //  - zero value -> len == 1 with limbs[0] == 0
+    limbs: []Limb,
+    len: usize,
+
+    const default_capacity = 4;
+
+    pub fn init(allocator: *Allocator) !Int {
+        return try Int.initCapacity(allocator, default_capacity);
+    }
+
+    pub fn initSet(allocator: *Allocator, value: var) !Int {
+        var s = try Int.init(allocator);
+        try s.set(value);
+        return s;
+    }
+
+    pub fn initCapacity(allocator: *Allocator, capacity: usize) !Int {
+        return Int{
+            .allocator = allocator,
+            .positive = true,
+            .limbs = block: {
+                var limbs = try allocator.alloc(Limb, math.max(default_capacity, capacity));
+                limbs[0] = 0;
+                break :block limbs;
+            },
+            .len = 1,
+        };
+    }
+
+    pub fn ensureCapacity(self: *Int, capacity: usize) !void {
+        if (capacity <= self.limbs.len) {
+            return;
+        }
+
+        self.limbs = try self.allocator.realloc(Limb, self.limbs, capacity);
+    }
+
+    pub fn deinit(self: *const Int) void {
+        self.allocator.free(self.limbs);
+    }
+
+    pub fn clone(other: *const Int) !Int {
+        return Int{
+            .allocator = other.allocator,
+            .positive = other.positive,
+            .limbs = block: {
+                var limbs = try other.allocator.alloc(Limb, other.len);
+                mem.copy(Limb, limbs[0..], other.limbs[0..other.len]);
+                break :block limbs;
+            },
+            .len = other.len,
+        };
+    }
+
+    pub fn copy(self: *Int, other: *const Int) !void {
+        if (self == other) {
+            return;
+        }
+
+        self.positive = other.positive;
+        try self.ensureCapacity(other.len);
+        mem.copy(Limb, self.limbs[0..], other.limbs[0..other.len]);
+        self.len = other.len;
+    }
+
+    pub fn swap(self: *Int, other: *Int) void {
+        mem.swap(Int, self, other);
+    }
+
+    pub fn dump(self: *const Int) void {
+        for (self.limbs) |limb| {
+            debug.warn("{x} ", limb);
+        }
+        debug.warn("\n");
+    }
+
+    pub fn negate(r: *Int) void {
+        r.positive = !r.positive;
+    }
+
+    pub fn abs(r: *Int) void {
+        r.positive = true;
+    }
+
+    pub fn isOdd(r: *const Int) bool {
+        return r.limbs[0] & 1 != 0;
+    }
+
+    pub fn isEven(r: *const Int) bool {
+        return !r.isOdd();
+    }
+
+    fn bitcount(self: *const Int) usize {
```

Comment:
Loosely, but did we want to expose the LLVM `ctpop` intrinsic?

Reply:
Yes, I believe that is #767